src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs

```diff
@@ -269,6 +269,27 @@ public override bool ComputeContainsGCPointers(DefType type)
             return someFieldContainsPointers;
         }
 
+        public override bool ComputeContainsByRefs(DefType type)
+        {
+            if (!type.IsByRefLike)
+                return false;
+
+            foreach (var field in type.GetFields())
+            {
+                if (field.IsStatic)
+                    continue;
+
+                TypeDesc fieldType = field.FieldType;
+                if (fieldType.IsByRef
+                    || ((fieldType is DefType df) && df.ContainsByRefs))
```

Comment:
Nit: we avoid `is` checks for DefType. These casts all involve deep walks of type hierarchy since it's an abstract type derived by another abstract type. ByRefLike type is guaranteed to be a `DefType`.

```suggestion
                    || (fieldType.IsByRefLike && ((DefType)fieldType).ContainsByRefs))
```